protocol/interop-access-list.md

```diff
@@ -0,0 +1,103 @@
+
+### Implementation example
+
+The `WARM_READ_THRESHOLD` value is up for debate, the cost difference between hot and cold is very large, but it would be good to have a consensus over it. The `calculateChecksum` will be expanded on the specs doc.
+
+```solidity
+uint256 internal constant WARM_READ_THRESHOLD = 150;
+
+function _isWarm(bytes32 _slot) internal view returns (bool isWarm, uint256 result) {
+    assembly {
+        let startGas := gas()
+        result := sload(_slot)
+        let endGas := gas()
+        isWarm := iszero(gt(sub(startGas, endGas), WARM_READ_COST))
+    }
+ }
+
+function validateMessage(Identifier calldata _id, bytes32 _msgHash) external {
+    bytes32 checksum = calculateChecksum(_id, _msgHash);
+
+    (bool _isSlotWarm,) = _isWarm(checksum);
+
+    if (!_isSlotWarm) revert NotWarm();
+
+    emit ExecutingMessage(_msgHash, _id);
+}
+```
+
+### Diagram
+
+Example using the `L2toL2CDM` with access-list in the transaction. The `L2toL2CDM` is not needed, it is being used for diagram clarity, any contract can call the `validateMessage` function.
+
+> Off-chain parts were skipped for simplification. The actual storage keys include more data not only the checksum. More info can be found on the specs doc.
+
+```mermaid
+sequenceDiagram
+    Relayer->>L2toL2CDM: tx with access-list: <br>{ "address": address(CrossL2Inbox), <br>"storageKeys": [checksum(Identifier, msgHash)] }
+    L2toL2CDM->>+CrossL2Inbox: validateMessage(Identifier, msgHash)
+    CrossL2Inbox-->>CrossL2Inbox: isWarm( checksum(Identifier, msgHash) )
+    Note over CrossL2Inbox: Gas cost < WARM_READ_COST
+    CrossL2Inbox-->>CrossL2Inbox: emit ExecutingMessage(msgHash, Identifier)
+    CrossL2Inbox->>L2toL2CDM: success
+```
+
+Example using the `L2toL2CDM` without access-list in the transaction. The `L2toL2CDM` is not needed, it is being used for diagram clarity, any contract can call the `validateMessage` function.
+
+```mermaid
+sequenceDiagram
+    Relayer->>L2toL2CDM: tx without access-list
+    L2toL2CDM->>+CrossL2Inbox: validateMessage(Identifier, msgHash)
+    CrossL2Inbox-->>CrossL2Inbox: isWarm( checksum(Identifier, msgHash) )
+    Note over CrossL2Inbox: Gas cost > WARM_READ_COST
+    CrossL2Inbox-->>L2toL2CDM: revert NotWarm()
+```
+
+### Alternatives considered
+
+Alternative solutions that kept the current dynamic declaration approach were explored. The most prominent one involved adding locking and unlocking functionality to the `CrossL2Inbox`. If the transaction referenced an invalid message, then that transaction would be included at the top of the next block when the `CrossL2Inbox` was locked. Because it was locked, the invalid transaction would revert. If it referenced an valid transaction, then it would be included when the `CrossL2Inbox` was unlocked. A special transaction took care of locking and unlocking the `CrossL2Inbox`.
+
+### Security considerations
+
+- Future changes in gas cost can affect this solution
```

Comment:
We may want to exand on this note that gas-introspection is fragile, and will need expert review with every EVM storage/call gas behavior change.
A custom opcode or precompile could do the access-list check in the future instead, to not rely on gas-introspection, if the gas-introspection is considered too problematic.